### ISSUE TITLE
Hotel Booking Demand

### ISSUE THREAD

**Issue by nimitshrestha** [nimitshrestha/nimitshrestha-DAML-2021-2022#2]
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv("/content/hotel_bookings.csv")
df
df.head(5)
df.describe()
df.info()
df.isnull().sum()

print("DataFrame Shape:", df.shape)
leadTime = df.loc[df.lead_time > 500, 'lead_time'] = 500
staysInWeekend = df.loc[df.stays_in_weekend_nights >=  5, 'stays_in_weekend_nights'] = 5
num_adult = df.loc[df.adults > 4, 'adults'] = 4
prevNotCancelled = df.loc[df.previous_bookings_not_canceled > 0, 'previous_bookings_not_canceled'] = 1
prevCancel = df.loc[df.previous_cancellations > 0, 'previous_cancellations'] = 1
staysInWeek = df.loc[df.stays_in_week_nights > 10, 'stays_in_week_nights'] = 10
bookinChanges = df.loc[df.booking_changes > 5, 'booking_changes'] = 5
num_Babies = df.loc[df.babies > 8, 'babies'] = 0
carParkingSpace = df.loc[df.required_car_parking_spaces > 5, 'required_car_parking_spaces'] = 0
num_children = df.loc[df.children > 8, 'children'] = 0
num_adr = df.loc[df.adr > 1000, 'adr'] = 1000

print("lead_time:",leadTime,"\n","stays_in_weekend_nights:",staysInWeekend,"\n",'adults:',num_adult,"\n",
      'previous_bookings_not_canceled:',prevNotCancelled,"\n",'previous_cancellations:',prevCancel,"\n",
      'stays_in_week_nights:',staysInWeek,"\n",'booking_changes:',bookinChanges,"\n",'babies:',num_Babies,"\n",
      'required_car_parking_spaces:',carParkingSpace,"\n",'children:',num_children,"\n",'adr:',num_adr)

print('Total Bookings cancelled')
print(df.is_canceled.value_counts())
print('Cancelation percentage')
print(df.is_canceled.value_counts(normalize=True))

plt.figure(figsize=(8,8))
#Canceled=1, Not canceled= 0
sns.countplot(df['is_canceled'], palette='flare')
plt.show()

df.country.value_counts(normalize=True)

df.arrival_date_month.value_counts(normalize=True)

plt.figure(figsize=(14,7))
sns.countplot(df['arrival_date_month'], palette='husl')
plt.show()

df.market_segment.value_counts(normalize=True)

plt.figure(figsize=(14,7))
sns.countplot(df['market_segment'], palette='crest')
plt.show()

df.meal.value_counts(normalize=True)

df.customer_type.value_counts(normalize=True)

plt.figure(figsize=(14,7))
sns.countplot(df['customer_type'], palette='flare')
plt.show()

plt.figure(figsize=(14,7))
sns.countplot(df['reserved_room_type'], palette='crest')
plt.show()

df.reservation_status.value_counts(normalize=True)

plt.figure(figsize=(12,5))
# Calculating average daily rate per person
df['adr_pp'] = df['adr'] / (df['adults'] + df['children']) 
actual_guests = df.loc[df["is_canceled"] != '0']
actual_guests['price'] = actual_guests['adr'] * (actual_guests['stays_in_weekend_nights'] + actual_guests['stays_in_week_nights'])
sns.lineplot(data = actual_guests, x = 'arrival_date_month', y = 'price', hue = 'hotel')
plt.show()

print('Most common countries:')
print(df.country.value_counts().head(3))
print('-'*50)
print('\n Most uncommon countries:')
print(df.country.value_counts().tail(3))

y=df['hotel']
x=df.drop(['hotel','arrival_date_month','meal','country','market_segment','distribution_channel','reserved_room_type','assigned_room_type',
           'reservation_status','reservation_status_date','customer_type','deposit_type','agent','company'],axis=1)
x_new = x.replace(np.nan, 0)

print(x_new)

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_new,y,random_state=1,test_size=0.2)

np.any(np.isnan(x_train))
np.all(np.isfinite(x_train))

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
rfc=RandomForestClassifier()
rfc.fit(x_train,y_train)

y_pred=rfc.predict(x_test)
conf =print(confusion_matrix(y_test, y_pred))

clf =print(classification_report(y_test, y_pred))
score=accuracy_score(y_test,y_pred)
print("ACCURACY OF THE MODEL IS:",score)
